### PR TITLE
ref(useHoverOverlay): refactor logic to avoid less state calls and optimize timers

### PR DESCRIPTION
**Pull request by JonasBa**
useHoverOverlay can be attached to a lot of elements on the sentry page. For example, when we render a lot of items in an assignee selector or a table with MxN rows, the instance count can easily grow to a couple hundred, which means we should be careful with the overhead we might be introducing. 

I initially looked into optimizing the performance of our assignee selector when I noticed we were firing off a couple hundred calls to clearTimeout when closing the dropdown (2x per instance for probably close to 150 people - this could be a much larger number for bigger orgs using sentry). As it turns out, these timers were never installed and clearing them wasn't doing anything, meaning we could just skip them instead.

Optimization applied: 
- removed calls to clear timeouts if those were never installed
- removed usage of `styled` - this bypasses the CSS generation from running through the emotion parser
- ~moved visible state value tracking from state to ref. Since the flip happens inside our timer and there are no reactive effects based off the visible state value, we can use a ref to track it, which avoids a state update and dirty checking of references~ I had to revert this as we return isOpen as a stateful value
- optimize the prop generation by writing to the initialized object as opposed to unnecessarily shallow cloning it

There are other optimization opportunities like tracking trigger elements through refs and a more efficient implementation of the isOverflown functionality, but I'm leaving it here for now.

### CODE REVIEW

File: static/app/components/assigneeSelectorDropdown.tsx [hunk 1]
```diff
@@ -619,21 +617,11 @@ const IconContainer = styled('div')`
...
-  disabled?: boolean;
-  py?: number;
-}>`
-  cursor: ${p => (p.disabled ? 'not-allowed' : 'pointer')};
+const MenuItemWrapper = styled('div')`
   display: flex;
   align-items: center;
   font-size: 13px;
   padding: ${space(0.5)} ${space(0.5)};
-  ${p =>
```
Comment (JonasBa) [inline static/app/components/assigneeSelectorDropdown.tsx:631]: this was never passed so I removed it

**Comment by JonasBa** [comment]
I was too quick on the ref change, there is also a `isOpen` so we need to rerender 😢

File: static/app/utils/useHoverOverlay.tsx [hunk 2]
```diff
@@ -102,57 +151,16 @@ function useHoverOverlay(
...
   }: UseHoverOverlayProps
 ) {
-  const [isVisible, setVisible] = useState(false);
-  const describeById = useMemo(() => domId(`${overlayType}-`), [overlayType]);
   const theme = useTheme();
+  const describeById = useMemo(() => domId(`${overlayType}-`), [overlayType]);
 
-  const isOpen = forceVisible ?? isVisible;
-
+  const [isOpen, setIsOpen] = useState(forceVisible ?? false);
```
Comment (gggritso) [inline static/app/utils/useHoverOverlay.tsx:157]: 🤔 doesn't this change the behaviour of `forceVisible`? It's only used as the initial value of the hook; if the user dismisses the overlay `isOpen` will be set to `false` which will hide the overlay even if `forceVisible` is still `true`, no?

**Review by gggritso — COMMENTED** [review]
Makes sense! Just one question about the visibility state change

Also,

> 2x per instance for probably close to 150 people - this could be a much larger number for bigger orgs using sentry

I think it's good to skip clearing the timeouts just for code hygeine, but out of curiosity, how long does it take to clear 300 non-existent timeouts?

**Comment by JonasBa** [comment]
> Makes sense! Just one question about the visibility state change
> 
> Also,
> 
> > 2x per instance for probably close to 150 people - this could be a much larger number for bigger orgs using sentry
> 
> I think it's good to skip clearing the timeouts just for code hygeine, but out of curiosity, how long does it take to clear 300 non-existent timeouts?

It took somewhere in the span of 3-5ms on my local profiles. It's not much, but keep in mind we often see these numbers in a 3-5x range in prod env (especially on non M1 macs :) ). If an org has a lot more users, which some of our larger orgs def do, cleaning up the timers alone can cause frame drops.

File: static/app/utils/useHoverOverlay.tsx (hunk 2 above)
Comment (JonasBa) [reply static/app/utils/useHoverOverlay.tsx:157]: No, I return `isOpen: forceVisible ?? isOpen` in the end, but I see how that is confusing. I'll find a better way

Comment (gggritso) [reply static/app/utils/useHoverOverlay.tsx:157]: Ah roger, I missed that part

**Review by gggritso — APPROVED** [review]
👍🏻 awesome!

File: static/app/utils/useHoverOverlay.tsx (hunk 2 above)
Comment (JonasBa) [reply static/app/utils/useHoverOverlay.tsx:157]: Yeah, I dont blame you. I will move it to what it was and use a ref to track the state inside the event hooks so we can avoid the invalidation

**Comment by JonasBa** [comment]
Not merging the PR yet as we just started collecting INP data and I would like to establish a baseline before adding optimizations

**Comment by JonasBa** [comment]
Error seems unrelated

**Pull request merged**